File: solutions/1401-1500/1497-check-if-array-pairs-are-divisible-by-k/solution.cpp
#include "solution.hpp"

namespace leetcode::p1497 {

bool solve(const std::vector<int>& arr, int k) {
    std::vector<int> remainderCounts(static_cast<std::size_t>(k), 0);

    for (int value : arr) {
        int remainder = ((value % k) + k) % k;
        ++remainderCounts[static_cast<std::size_t>(remainder)];
    }

    if (remainderCounts[0] % 2 != 0) {
        return false;
    }

    for (int remainder = 1; remainder <= k / 2; ++remainder) {
        if (remainderCounts[static_cast<std::size_t>(remainder)] !=
            remainderCounts[static_cast<std::size_t>(k - remainder)]) {
            return false;
        }
    }

    return true;
}

} // namespace leetcode::p1497

File: solutions/1401-1500/1497-check-if-array-pairs-are-divisible-by-k/solution.ml
let solve arr k =
  let remainder_counts = Array.make k 0 in
  List.iter
    (fun n ->
      let remainder = ((n mod k) + k) mod k in
      remainder_counts.(remainder) <- remainder_counts.(remainder) + 1)
    arr;
  if remainder_counts.(0) mod 2 <> 0 then false
  else begin
    let balanced = ref true in
    for remainder = 1 to (k - 1) / 2 do
      if remainder_counts.(remainder) <> remainder_counts.(k - remainder) then
        balanced := false
    done;
    if k mod 2 = 0 && remainder_counts.(k / 2) mod 2 <> 0 then balanced := false;
    !balanced
  end
